streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -220,11 +237,27 @@ public void setStateListener(final KafkaStreams.StateListener listener) {
         stateListener = listener;
     }
 
+    /**
+     * Sets the state
+     * @param newState New state
+     */
     private void setState(final State newState) {
+
         synchronized (stateLock) {
+
+            // there are cases when we shouldn't check if a transition is valid, e.g.,
+            // when, for testing, Kafka Streams is closed multiple times. We could either
+            // check here and immediately return for those cases, or add them to the transition
+            // diagram (but then the diagram would be confusing and have transitions like
+            // NOT_RUNNING->NOT_RUNNING).
+            if (newState != NOT_RUNNING && (state == State.NOT_RUNNING || state == PENDING_SHUTDOWN)) {
```

Comment:
Yes. There are cases when the instance is closing, but one thread just moved to RUNNING. Once the instance is in pending shutdown, it should only go to not running, should never change back to running.